### PR TITLE
Dynamic sync of ACL based on incoming token and user

### PR DESCRIPTION
**Pull request by jcantrill**

### CODE REVIEW

File: src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	private final String caCertFile;
+	private final Boolean trustCerts;
+
+	public DynamicACLFilter(final Settings settings, final Client client, final ESLogger logger){
+		this.logger = logger;
+		this.esClient = client;
+		this.proxyUserHeader = settings.get(SEARCHGUARD_AUTHENTICATION_PROXY_HEADER, DEFAULT_AUTH_PROXY_HEADER);
+		this.searchGuardIndex = settings.get(SEARCHGUARD_CONFIG_INDEX_NAME, DEFAULT_SECURITY_CONFIG_INDEX);
+		this.masterApiUrl = Utils.getSystemPropertyOrEnvVar(MASTER_API_URL);
+		this.caCertFile = Utils.getSystemPropertyOrEnvVar(CERT_CA_FILE, "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt");
```
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java]: This is redundant - client handles that for you.

File: src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		this.trustCerts = Utils.getSystemPropertyOrEnvVar(TRUST_CERTS, Boolean.FALSE);
```
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java]: Ditto.

File: src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		if (auth.length >= 2 && "Bearer".equals(auth[0])) {
+			return auth[1];
+		}
+		return "";
+	}
+	
+
+	private void updateCache(final String user, final String token) {
+		logger.debug("Updating the cache for user '{}'", user);
+		OpenShiftClient osClient = null;
```
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java]: Change to try-with-resources style - more readable, client is autocloseable.

**Comment by jimmidyson** [comment]
I do have concerns with how this will work with a cluster - is this going to introduce conflicts to the ACL?

File: src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	private final String masterApiUrl;
+	private final String caCertFile;
+	private final Boolean trustCerts;
+
+	public DynamicACLFilter(final Settings settings, final Client client, final ESLogger logger){
+		this.logger = logger;
+		this.esClient = client;
+		this.proxyUserHeader = settings.get(SEARCHGUARD_AUTHENTICATION_PROXY_HEADER, DEFAULT_AUTH_PROXY_HEADER);
+		this.searchGuardIndex = settings.get(SEARCHGUARD_CONFIG_INDEX_NAME, DEFAULT_SECURITY_CONFIG_INDEX);
+		this.masterApiUrl = Utils.getSystemPropertyOrEnvVar(MASTER_API_URL);
```
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java]: Client uses `KUBERNETES_MASTER` env var & defaults to `kubernetes.default.svc` so it works in-cluster without any config. Remove.

File: src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java (hunk 2 above)
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/plugin/DynamicACLFilter.java]: Remove please.

File: src/main/java/io/fabric8/elasticsearch/plugin/osclient/Fabric8OpenShiftClientAdapter.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import io.fabric8.kubernetes.client.ConfigBuilder;
+import io.fabric8.openshift.api.model.Project;
+import io.fabric8.openshift.client.DefaultOpenshiftClient;
+
+public class Fabric8OpenShiftClientAdapter implements OpenShiftClient {
```
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/plugin/osclient/Fabric8OpenShiftClientAdapter.java]: Why do we need this adapter?

File: src/main/java/io/fabric8/elasticsearch/plugin/osclient/OpenShiftClient.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+package io.fabric8.elasticsearch.plugin.osclient;
+
+import java.util.Set;
+
+/**
+ * Abstraction to the OpenShift api server
+ * @author jeff.cantrill
+ *
+ */
+public interface OpenShiftClient {
```
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/plugin/osclient/OpenShiftClient.java]: This interface seems a bit redundant to me. I'd remove personally & just use the `client.projects().list()` call when needed.

**Comment by jcantrill** [comment]
@jimmidyson what do you mean by 'conflicts to the ACL'  @sosiouxme discussed our issues with doing dynamic ACLs and for this to work we believe each member will have to keep its own copy of the ACL unless @ewolinetz determines we can do othewise.  Please review.  added ACL expiration as well.

**Comment by jimmidyson** [comment]
Yeah if we use separate acl for each node that's OK. I was worried about concurrent updates to the same acl doc.

File: src/main/java/io/fabric8/elasticsearch/plugin/osclient/Fabric8OpenShiftClientAdapter.java (hunk 5 above)
Comment (jcantrill) [inline src/main/java/io/fabric8/elasticsearch/plugin/osclient/Fabric8OpenShiftClientAdapter.java]: Originally added adapter when I was trying to make it work and pulled in the openshift-restclient-java.  This was before project retrieval was supported.  I'll remove it.

**Comment by jcantrill** [comment]
@jimmidyson Updated to remove interface to fabric8 openshift client.  @ewolinetz added readme with info about running from eclipse.  PTAL

**Comment by jimmidyson** [comment]
:+1:

**Pull request merged**